File: code/prime_factorize.py
from math import gcd
def pollard(n):
	f = lambda x: x * x % n + 1
	x, y, t, prd, i = 0, 0, 30, 2, 1
	while t % 40 or gcd(prd, n) == 1:
		if x == y:
			i += 1
			x, y = i, f(i)
		if q := prd * (max(x,y) - min(x,y)) % n:
			prd = q
		x, y = f(x), f(f(y))
		t += 1
	return gcd(prd, n)
def factor(n):
	if n == 1: return []
	if isPrime(n): return [n]
	x = pollard(n)
	return factor(x) + factor(n // x)

File: code/prime_factorize.cpp
#include "is_prime.cpp"
ull pollard(ull n) {
	auto f = [n](ull x) {return modmul(x, x, n)+1;};
	ull x = 0, y = 0, t = 30, prd = 2, i = 1;
	while (t++ % 40 || gcd(prd, n) == 1) {
		if (x == y) x = ++i, y = f(x);
		ull q = modmul(prd, max(x,y)-min(x,y), n);
		if (q) prd = q;
		x = f(x), y = f(f(y));
	}
	return gcd(prd, n);
}
vector<ull> factor(ull n) {
	if (n == 1) return {};
	if (isPrime(n)) return {n};
	ull x = pollard(n);
	auto l = factor(x), r = factor(n / x);
	l.insert(l.end(), all(r));
	return l;
}
